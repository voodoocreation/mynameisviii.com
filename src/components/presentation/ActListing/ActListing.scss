@import "../../../scss/variables", "../../../scss/mixins";

.ActListing {
  padding: $spacing-half;

  h3 {
    margin: 0 0 $spacing-half;
  }

  a {
    display: flex;
    flex-direction: row-reverse;
    align-items: center;
  }

  img {
    border-radius: 100%;
    overflow: hidden;
  }

  &--details {
    display: flex;
    flex-flow: row wrap;
    flex-basis: 85%;
    align-content: center;
    padding-left: $spacing-half;

    > * {
      flex-basis: 100%;
    }
  }

  .Meta {
    display: flex;
  }

  &--image {
    flex-basis: 15%;
  }

  @include when-above($screen-md-min) {
    &--details {
      flex-basis: 70%;
    }

    &--image {
      flex-basis: 30%;
    }
  }
}
