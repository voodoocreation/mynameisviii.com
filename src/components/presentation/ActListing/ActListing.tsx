import cn from "classnames";
import * as React from "react";
import { MdLocationOn, MdMusicNote } from "react-icons/md";

import Link from "../Link/Link";
import Meta from "../Meta/Meta";

interface IState {
  isRendered: boolean;
}

interface IProps extends IPerformer {
  onLoad?: (name: string) => void;
}

export default class ActListing extends React.Component<IProps, IState> {
  public readonly state = {
    isRendered: false
  };

  private imageRef: React.RefObject<HTMLImageElement> = React.createRef();

  public componentDidMount() {
    if (this.imageRef.current && this.imageRef.current.complete) {
      setTimeout(this.onLoad, 1);
    }
  }

  public render() {
    const { onLoad, ...act } = this.props;
    const { isRendered } = this.state;

    return (
      <article className={cn("ActListing", { isRendered })}>
        <Link className="ActListing-link" href={act.url} isExternal={true}>
          <div className="ActListing-details">
            <h3>{act.name}</h3>

            <section className="ActListing-meta">
              <Meta
                className="ActListing-genre"
                icon={<MdMusicNote />}
                labelConstant="GENRE"
              >
                {act.genre}
              </Meta>

              <Meta
                className="ActListing-location"
                icon={<MdLocationOn />}
                labelConstant="LOCATION"
              >
                {act.location.name}
              </Meta>
            </section>
          </div>
          <figure className="ActListing-image">
            <img
              src={act.imageUrl}
              alt={act.name}
              ref={this.imageRef}
              onLoad={this.onLoad}
              onError={this.onLoad}
            />
          </figure>
        </Link>
      </article>
    );
  }

  private onLoad = () => {
    this.setState({
      isRendered: true
    });

    if (this.props.onLoad) {
      this.props.onLoad(this.props.name);
    }
  };
}
