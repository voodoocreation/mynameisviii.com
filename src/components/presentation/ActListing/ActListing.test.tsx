import { render, shallow } from "enzyme";
import * as React from "react";

import ActListing from "./ActListing";

const g: any = global;

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    genre: "genre",
    imageUrl: "imageUrl",
    location: {
      address: "location address",
      name: "location name",
      type: "location type",
      url: "location url"
    },
    name: "name",
    type: "type",
    url: "url",
    ...fromTestProps
  };

  return {
    actual: fn(<ActListing {...props} />),
    props
  };
};

describe("[components] <ActListing />", () => {
  it("renders correctly", () => {
    const { actual } = setup(render);
    expect(actual).toMatchSnapshot();
  });

  it("updates `isRendered` state after image has loaded", () => {
    const { actual } = setup(shallow);

    actual.find("img").simulate("load");
    expect(actual.state("isRendered")).toBe(true);
    expect(actual.render()).toMatchSnapshot();
  });

  it("triggers onLoad prop after image has loaded", () => {
    const { actual, props } = setup(shallow, {
      onLoad: jest.fn()
    });

    actual.find("img").simulate("load");
    expect(props.onLoad).toHaveBeenCalled();
  });

  it("updates `isRendered` state and triggers onLoad prop when image has previously been loaded", () => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: true
    });

    const { actual, props } = setup(shallow, {
      onLoad: jest.fn()
    });

    actual.find("img").simulate("load");
    expect(props.onLoad).toHaveBeenCalled();
    expect(actual.state("isRendered")).toBe(true);
    expect(actual.render()).toMatchSnapshot();
  });
});
