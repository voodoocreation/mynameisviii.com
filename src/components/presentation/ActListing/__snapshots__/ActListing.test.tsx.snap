// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <ActListing /> renders correctly 1`] = `
<article
  class="ActListing"
>
  <a
    class="ActListing-link"
    href="url"
    rel="noopener"
    target="_blank"
  >
    <div
      class="ActListing-details"
    >
      <h3>
        name
      </h3>
      <section
        class="ActListing-meta"
      >
        <div
          class="Meta ActListing-genre"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Genre
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            genre
          </span>
        </div>
        <div
          class="Meta ActListing-location"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Location
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            location name
          </span>
        </div>
      </section>
    </div>
    <figure
      class="ActListing-image"
    >
      <img
        alt="name"
        src="imageUrl"
      />
    </figure>
  </a>
</article>
`;

exports[`[presentation] <ActListing /> updates \`isRendered\` state after image has loaded 1`] = `
<article
  class="ActListing isRendered"
>
  <a
    class="ActListing-link"
    href="url"
    rel="noopener"
    target="_blank"
  >
    <div
      class="ActListing-details"
    >
      <h3>
        name
      </h3>
      <section
        class="ActListing-meta"
      >
        <div
          class="Meta ActListing-genre"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Genre
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            genre
          </span>
        </div>
        <div
          class="Meta ActListing-location"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Location
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            location name
          </span>
        </div>
      </section>
    </div>
    <figure
      class="ActListing-image"
    >
      <img
        alt="name"
        src="imageUrl"
      />
    </figure>
  </a>
</article>
`;

exports[`[presentation] <ActListing /> updates \`isRendered\` state and triggers onLoad prop when image has previously been loaded 1`] = `
<article
  class="ActListing isRendered"
>
  <a
    class="ActListing-link"
    href="url"
    rel="noopener"
    target="_blank"
  >
    <div
      class="ActListing-details"
    >
      <h3>
        name
      </h3>
      <section
        class="ActListing-meta"
      >
        <div
          class="Meta ActListing-genre"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Genre
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            genre
          </span>
        </div>
        <div
          class="Meta ActListing-location"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
           
          <span
            class="Meta-label"
          >
            <span>
              Location
            </span>
            :
          </span>
           
          <span
            class="Meta-value"
          >
            location name
          </span>
        </div>
      </section>
    </div>
    <figure
      class="ActListing-image"
    >
      <img
        alt="name"
        src="imageUrl"
      />
    </figure>
  </a>
</article>
`;
